### ISSUE TITLE
[BUG] BetterFPS - Game Crash 

### ISSUE THREAD

**Issue by xedis** [DysonSphereMod/QOL#82]
Attached is my game log from %APPDATA%\LocalLow\YouthCatStudios\
I turned on a bunch of BepinEx logging options prior to this crash
Crash occurs on line 34188
[output_log.txt](https://github.com/DysonSphereMod/QOL/files/6329705/output_log.txt)
I was pulling a belt from an ILS at the time, I didnt notice my research time but am loading game again. 

PS. In future I'll trim the log so it only includes the relevant loaded game instance 

Edit* - Research- 
========== OUTPUTTING STACK TRACE ==================

0x00007FF923ED0F40 (UnityPlayer) UnityMain
0x00007FF923ED1FFF (UnityPlayer) UnityMain
0x00007FF923EC760E (UnityPlayer) UnityMain
0x00007FF923EC41E3 (UnityPlayer) UnityMain
0x00007FF923EE389B (UnityPlayer) UnityMain
0x000000001EF8B941 (Mono JIT Code) (wrapper managed-to-native) UnityEngine.TextGenerator:Populate_Internal_Injected (string,UnityEngine.Font,UnityEngine.Color&,int,single,single,UnityEngine.FontStyle,bool,bool,int,int,int,int,bool,UnityEngine.TextAnchor,single,single,single,single,bool,bool,uint&)
0x000000001EF8B778 (Mono JIT Code) UnityEngine.TextGenerator:Populate_Internal (string,UnityEngine.Font,UnityEngine.Color,int,single,single,UnityEngine.FontStyle,bool,bool,int,int,int,int,bool,UnityEngine.TextAnchor,single,single,single,single,bool,bool,uint&)
0x000000001EF8B5F8 (Mono JIT Code) UnityEngine.TextGenerator:Populate_Internal (string,UnityEngine.Font,UnityEngine.Color,int,single,single,UnityEngine.FontStyle,bool,bool,int,int,UnityEngine.VerticalWrapMode,UnityEngine.HorizontalWrapMode,bool,UnityEngine.TextAnchor,UnityEngine.Vector2,UnityEngine.Vector2,bool,bool,UnityEngine.TextGenerationError&)
0x000000001EF8B157 (Mono JIT Code) UnityEngine.TextGenerator:PopulateAlways (string,UnityEngine.TextGenerationSettings)
0x000000001EF8AEBE (Mono JIT Code) UnityEngine.TextGenerator:PopulateWithError (string,UnityEngine.TextGenerationSettings)
0x0000000106DCA4F4 (Mono JIT Code) UnityEngine.TextGenerator:Populate (string,UnityEngine.TextGenerationSettings)
0x0000000106DCA3D4 (Mono JIT Code) UnityEngine.TextGenerator:GetPreferredHeight (string,UnityEngine.TextGenerationSettings)
0x0000000106DCA1D6 (Mono JIT Code) UnityEngine.UI.Text:get_preferredHeight ()
0x0000000106E6DF41 (Mono JIT Code) UIResearchResultWindow:SetTechId (int)
0x000000015418FE29 (Mono JIT Code) UIGeneralTips:OnTechUnlocked (int,int)
0x000000015418FB06 (Mono JIT Code) GameHistoryData:NotifyTechUnlock (int,int)
0x000000001C02C232 (Mono JIT Code) (wrapper dynamic-method) FactorySystem:DMD<FactorySystem::GameTick> (FactorySystem,long,bool)
0x0000000038A25186 (Mono JIT Code) PlanetFactory:GameTick (long)
0x0000000038A20CAC (Mono JIT Code) BetterFPS.BetterFPS/<>c__DisplayClass6_1:<GameData_GameTick_Patch>b__1 ()
0x0000000038A20265 (Mono JIT Code) System.Threading.Tasks.Task:InnerInvoke ()
0x0000000038A1F34C (Mono JIT Code) System.Threading.Tasks.Task:Execute ()
0x0000000038A1E765 (Mono JIT Code) System.Threading.Tasks.Task:ExecutionContextCallback (object)
0x0000000038A18FB8 (Mono JIT Code) System.Security.SecurityContext:Run (System.Security.SecurityContext,System.Threading.ContextCallback,object)
0x0000000038A17B76 (Mono JIT Code) System.Threading.ExecutionContext:Run (System.Threading.ExecutionContext,System.Threading.ContextCallback,object)
0x0000000038A16F1A (Mono JIT Code) System.Threading.Tasks.Task:ExecuteWithThreadLocal (System.Threading.Tasks.Task&)
0x0000000038A16489 (Mono JIT Code) System.Threading.Tasks.Task:ExecuteEntry (bool)
0x0000000038A162B1 (Mono JIT Code) System.Threading.Tasks.ThreadPoolTaskScheduler:TaskExecuteWaitCallback (object)
0x0000000006370126 (Mono JIT Code) (wrapper runtime-invoke) object:runtime_invoke_void__this___object (object,intptr,intptr,intptr)
0x00007FF95ACA6783 (mono) mono_set_defaults
0x00007FF95ABF8AD9 (mono) mono_runtime_invoke
0x00007FF95ABFEFEF (mono) mono_runtime_invoke_array
0x00007FF95ABFF2C4 (mono) mono_runtime_invoke_array
0x00007FF95AC1F95B (mono) mono_security_set_mode
0x00007FF95AC203B2 (mono) mono_security_set_mode
0x00007FF95AC240FF (mono) mono_thread_interruption_request_flag
0x00007FF95ACDA1AB (mono) mono_unity_gc_is_disabled
0x00007FF98B427034 (KERNEL32) BaseThreadInitThunk
0x00007FF98D202651 (ntdll) RtlUserThreadStart

========== END OF STACKTRACE ===========

**Comment by xedis** [comment]
[output_log-2nd crash.txt](https://github.com/DysonSphereMod/QOL/files/6329879/output_log-2nd.crash.txt)
========== OUTPUTTING STACK TRACE ==================

0x00007FF913BEFF6E (UnityPlayer) UnityMain
0x00007FF913CEFFD9 (UnityPlayer) UnityMain
0x0000000017AF4E81 (Mono JIT Code) (wrapper managed-to-native) UnityEngine.ComputeBuffer:InternalSetData (System.Array,int,int,int,int)
0x00000000400369DF (Mono JIT Code) UnityEngine.ComputeBuffer:SetData (System.Array)
0x000000010AAFE4EB (Mono JIT Code) UIStorageGrid:OnStorageContentChanged ()
0x0000000040066AD2 (Mono JIT Code) StorageComponent:NotifyStorageChange ()
0x000000010AA35B40 (Mono JIT Code) StorageComponent:AddItem (int,int,bool)
0x00000000400FDE13 (Mono JIT Code) PlanetFactory:InsertInto (int,int,int)
0x00000000400FD6CC (Mono JIT Code) InserterComponent:InternalUpdate (PlanetFactory,int[][],AnimData[],single)
0x00000000400FD11D (Mono JIT Code) FactorySystem:GameTickInserters (long,bool)
0x00000000400ED6CE (Mono JIT Code) PlanetFactory:GameTick (long)
0x00000000400E8D6C (Mono JIT Code) BetterFPS.BetterFPS/<>c__DisplayClass6_1:<GameData_GameTick_Patch>b__1 ()
0x00000000400E86B5 (Mono JIT Code) System.Threading.Tasks.Task:InnerInvoke ()
0x00000000400E76BC (Mono JIT Code) System.Threading.Tasks.Task:Execute ()
0x00000000400E6995 (Mono JIT Code) System.Threading.Tasks.Task:ExecutionContextCallback (object)
0x00000000400E1748 (Mono JIT Code) System.Security.SecurityContext:Run (System.Security.SecurityContext,System.Threading.ContextCallback,object)
0x00000000400E0756 (Mono JIT Code) System.Threading.ExecutionContext:Run (System.Threading.ExecutionContext,System.Threading.ContextCallback,object)
0x00000000400DFA9A (Mono JIT Code) System.Threading.Tasks.Task:ExecuteWithThreadLocal (System.Threading.Tasks.Task&)
0x00000000400DF0B9 (Mono JIT Code) System.Threading.Tasks.Task:ExecuteEntry (bool)
0x00000000400DEEE1 (Mono JIT Code) System.Threading.Tasks.ThreadPoolTaskScheduler:TaskExecuteWaitCallback (object)
0x00000000063A0126 (Mono JIT Code) (wrapper runtime-invoke) object:runtime_invoke_void__this___object (object,intptr,intptr,intptr)
0x00007FF925E66783 (mono) mono_set_defaults
0x00007FF925DB8AD9 (mono) mono_runtime_invoke
0x00007FF925DBEFEF (mono) mono_runtime_invoke_array
0x00007FF925DBF2C4 (mono) mono_runtime_invoke_array
0x00007FF925DDF95B (mono) mono_security_set_mode
0x00007FF925DE03B2 (mono) mono_security_set_mode
0x00007FF925DE40FF (mono) mono_thread_interruption_request_flag
0x00007FF925E9A1AB (mono) mono_unity_gc_is_disabled
0x00007FF98B427034 (KERNEL32) BaseThreadInitThunk
0x00007FF98D202651 (ntdll) RtlUserThreadStart

========== END OF STACKTRACE ===========

**Comment by Simon-Gau** [comment]
I got the exact same crash and output log after playing for a decent time, placing belts and all, no research going. Wasn't pulling a belt out a logistics station, but I previously did and it worked.  Loaded the game, went back to building, and crashed in the same way with the same log when I placed a power satellite thingy.

No research was running.

**Comment by markkujarvinen** [comment]
The code in GameData_GameTick_Patch copies factory array and then works on it in threads, as it looks to be array on the original code could the multithread part just create ConcurrentQueue of the index numbers and use the original array to reference the factories, then the null-check should be sufficient like it is in the "normal" side of the code. Obviously this still leaves race condition on the active GameTick but as the normal game does not suffer from such it might be safe.

**Comment by xedis** [comment]
Here's a crash from accessing a storage container, a big one. No storage related mods installed.

[output_log.txt](https://github.com/DysonSphereMod/QOL/files/6330624/output_log.txt)

**Comment by glouel** [comment]
I experienced the exact same crash with 1.0, while picking all the belts from a medium storage chest. 

Just tried 1.0.1 and the issue remains fyi: 

```
========== OUTPUTTING STACK TRACE ==================

0x00007FF86C84FF6E (UnityPlayer) UnityMain
0x00007FF86C94FFD9 (UnityPlayer) UnityMain
0x000002764DC0A541 (Mono JIT Code) (wrapper managed-to-native) UnityEngine.ComputeBuffer:InternalSetData (System.Array,int,int,int,int)
0x000002764DC6DA4F (Mono JIT Code) UnityEngine.ComputeBuffer:SetData (System.Array)
0x000002770C0BA270 (Mono JIT Code) UIStorageGrid:OnStorageContentChanged ()
0x00000276E474D642 (Mono JIT Code) StorageComponent:NotifyStorageChange ()
0x0000027749F9C28D (Mono JIT Code) StorageComponent:AddItem (int,int,bool)
0x00000276E47EC643 (Mono JIT Code) PlanetFactory:InsertInto (int,int,int)
0x00000276E47EBEFC (Mono JIT Code) InserterComponent:InternalUpdate (PlanetFactory,int[][],AnimData[],single)
0x00000276E47EB90D (Mono JIT Code) FactorySystem:GameTickInserters (long,bool)
0x00000276E47DC76E (Mono JIT Code) PlanetFactory:GameTick (long)
0x00000276E47D8F4C (Mono JIT Code) BetterFPS.BetterFPS/<>c__DisplayClass6_1:<GameData_GameTick_Patch>b__1 ()
0x00000276E47D872B (Mono JIT Code) System.Threading.Tasks.Task:InnerInvoke ()
0x00000276E47D7CFC (Mono JIT Code) System.Threading.Tasks.Task:Execute ()
0x00000276E47D777B (Mono JIT Code) System.Threading.Tasks.Task:ExecutionContextCallback (object)
0x00000276E47D20C2 (Mono JIT Code) System.Security.SecurityContext:Run (System.Security.SecurityContext,System.Threading.ContextCallback,object)
0x00000276E47D0E26 (Mono JIT Code) System.Threading.ExecutionContext:Run (System.Threading.ExecutionContext,System.Threading.ContextCallback,object)
0x00000276E47D055F (Mono JIT Code) System.Threading.Tasks.Task:ExecuteWithThreadLocal (System.Threading.Tasks.Task&)
0x00000276E47CFD79 (Mono JIT Code) System.Threading.Tasks.Task:ExecuteEntry (bool)
0x00000276E47CFC18 (Mono JIT Code) System.Threading.Tasks.ThreadPoolTaskScheduler:TaskExecuteWaitCallback (object)
0x0000027632070136 (Mono JIT Code) (wrapper runtime-invoke) object:runtime_invoke_void__this___object (object,intptr,intptr,intptr)
0x00007FF879D66783 (mono) mono_set_defaults
0x00007FF879CB8AD9 (mono) mono_runtime_invoke
0x00007FF879CBEFEF (mono) mono_runtime_invoke_array
0x00007FF879CBF2C4 (mono) mono_runtime_invoke_array
0x00007FF879CDF95B (mono) mono_security_set_mode
0x00007FF879CE03B2 (mono) mono_security_set_mode
0x00007FF879CE40FF (mono) mono_thread_interruption_request_flag
0x00007FF879D9A1AB (mono) mono_unity_gc_is_disabled
0x00007FF8BECF7034 (KERNEL32) BaseThreadInitThunk
0x00007FF8BF6FD241 (ntdll) RtlUserThreadStart

========== END OF STACKTRACE ===========
```

The crash is 100% reproducible here, if you need the save file to investigate. 

(and many thanks for all your hard work on this, you're really making the game a lot better to play)

**Comment by brokenmass** [comment]
fixed in betterFPS 1.0.2

**Comment by glouel** [comment]
(Just confirming, it's indeed fixed for me and no longer reproduce with 1.0.2, thanks again !)

**Comment by markkujarvinen** [comment]
I looked the code and in the NotifyStorages you are clearing in the end the unlockedTechs and not the changedStorages which you loop through, so a copy-paste error looks like.

**Comment by brokenmass** [comment]
@markkujarvinen yes i noticed that too. it's already fixed in 1.0.3 (that i hadn't committed until some minutes ago_
